cip/1.accepted/CIP2021-07-07-Grouping-keys-and-aggregation-expressions.adoc

```diff
@@ -0,0 +1,779 @@
+
+This proposal provides a such precise semantics.
+
+== Proposal
+
+=== Syntax
+
+This proposal does not propose any net-new syntax.
+
+=== Semantics
+
+The proposed grouping and aggregation semantics is defined as a rewrite to the baseline semantics (similar to <<Simple rewrite semantics>> discussed above).
+The proposed semantics does not cover all syntactically possible queries and hence requires a syntax restriction to prohibit queries that are not covered.
+We discuss the <<Rewrite>> and the <<Syntax restriction>> in the follow two subsections.
+We simplify this discussion by ignoring row ordering and pagination as well as omitted aliases.
+Subsequently, we give a separate and brief consideration of how to these aspects fit into the proposal, cf. <<Row ordering and pagination>> and <<Omitted aliases>>, respectively.
+
+==== Rewrite
+
+For an expression _ex_, let _AGG_(_ex_) be the set of (sub-)expressions _aggex_ of the form _agg_(_preEx_).
+
+For a <ProjectionItem> _p_ = `_postEx_ AS _al_`, let _AGG_(_p_) be the set of (sub-)expressions _aggex_ of the form _agg_(_preEx_), i.e. _AGG_(_p_) = _AGG_(_postEx_).
+
+The <ProjectionItem>s in _L_ is split according to _AGG_(_p_) in two cases
+
+* <ProjectionItem>s _p_ in _L_ is an aggregate if _AGG_(_p_) is non-empty
+* <ProjectionItem>s _p_ in _L_ is a grouping key if _AGG_(_p_) is empty
+
+For a <ProjectionItem> _p_ = `_ex_ AS _al_`,
+
+* If _AGG_(_p_) = ∅
+** Let `_PRE_PROJ_(_p_)` be `_ex_ AS _al_`,
+** Let `_AGGR_(_p_)` be `_al_ AS _al_`, and
+** Let `_POST_PROJ_(_p_)` be `_al_ AS _al_`
+* If _AGG_(_p_) = {`_agg_~1~(_preEx_~1~)`, `_agg_~2~(_preEx_~2~)`, ..., `_agg~N~_(_preEx~N~_)`} with _N_ > 0
+** Let `_PRE_PROJ_(_p_)` be `_preEx_~1~ AS _al_+++_+++1, _preEx_~2~ AS _al_+++_+++2, ..., _preEx~N~_ AS _al_+++_+++_N_`,
+** Let `_AGGR_(_p_)` be `_agg_~1~(_al_+++_+++1) AS _al_+++_+++1, _agg_~2~(_al_+++_+++2) AS _al_+++_+++2, ..., _agg~N~_(_al_+++_+++_N_) AS _al_+++_+++_N_`, and
+** Let `_POST_PROJ_(_p_)` be `_postEx_ AS _al_` where _postEx_ is _ex_ with each `_agg~i~_(_preEx~i~_)` in _AGG_(_p_) being replaced by `_al_+++_+++_i_` for 1 ≤ _i_ ≤ _N_.
+
+[IMPORTANT]
+.Rewrite semantics
+====
+`RETURN _p_~1~, _p_~2~, ..., _p~N~_` is effectively equivalent to
+
+[source, cypher, subs="quotes"]
+----
+WITH _PRE_PROJ_(_p_~1~), _PRE_PROJ_(_p_~2~), ..., _PRE_PROJ_(_p~N~_)
+WITH _AGGR_(_p_~1~), _AGGR_(_p_~2~), ..., _AGGR_(_p~N~_)
+RETURN _POST_PROJ_(_p_~1~), _POST_PROJ_(_p_~2~), ..., _POST_PROJ_(_p~N~_)
+----
+
+Analogously, `WITH _p_~1~, _p_~2~, ..., _p~N~_` is effectively equivalent to
+
+[source, cypher, subs="quotes"]
+----
+WITH _PRE_PROJ_(_p_~1~), _PRE_PROJ_(_p_~2~), ..., _PRE_PROJ_(_p~N~_)
+WITH _AGGR_(_p_~1~), _AGGR_(_p_~2~), ..., _AGGR_(_p~N~_)
+WITH _POST_PROJ_(_p_~1~), _POST_PROJ_(_p_~2~), ..., _POST_PROJ_(_p~N~_)
+----
+====
+
+.Rewrite semantics
+====
+The RETURN clause in <<Q4>>
+
+[source, cypher]
+----
+RETURN a AS a, (a + SUM(b*c) - MIN(c)) * 2 AS foo
```

Comment:
I would have liked to see an example where the expression that contains an aggregation also contains a  non-leaf grouping expression. For example:

```
RETURN a.x, a.x + MIN(a.y)
```

How would this be rewritten with the rules?

Reply:
```
RETURN a.x AS foo, a.x + MIN(a.y) AS bar
```

what it should become

```
WITH a.x AS foo, a.y AS bar_1
WITH foo AS foo, MIN(bar_1) AS bar_1
RETURN foo AS foo, foo + bar_1  AS foo
```

You have a good point, in that currently the rewrite does not do that yet.

But it seems easy to fix. The definition of _POST_PROJ_(_p_) needs adjust, so that is also replaces all subexpression that are grouping keys with their alias.